### PR TITLE
Cherry-pick our workaround patches

### PR DESCRIPTION
**Pull request by nre-ableton**
This PR basically just brings the `ableton-master` to the same point as the old `nre_master_ableton-octo` branch (which lives on under the tag name `ableton-initial-migration`). The only difference is that I've cleaned up the commit messages a bit.

ping @ala-ableton

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,6 +4,7 @@ project(Cucumber-Cpp)
 
 set(CUKE_USE_STATIC_BOOST ${WIN32} CACHE BOOL "Statically link Boost (except boost::test)")
 set(CUKE_USE_STATIC_GTEST ON CACHE BOOL "Statically link Google Test")
+set(CUKE_ENABLE_SHARED_LIB OFF CACHE BOOL "Generate a shared library")
```
Comment (ala-ableton) [inline CMakeLists.txt]: I don't think we need to change the `CMakeLists.txt` in this commit.

File: include/cucumber-cpp/internal/CukeEngineImpl.hpp [hunk 2]
```diff
@@ -13,7 +13,7 @@ namespace internal {
  * Currently it is a wrapper around CukeCommands. It will have its own
  * implementation when feature #31 is complete.
  */
-class CukeEngineImpl : public CukeEngine {
+class CUKE_API_ CukeEngineImpl : public CukeEngine {
```
Comment (ala-ableton) [inline include/cucumber-cpp/internal/CukeEngineImpl.hpp]: Missing `#include "CukeDll.hpp"`

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -33,8 +33,14 @@ if(CMAKE_EXTRA_GENERATOR OR MSVC_IDE)
...
-add_library(cucumber-cpp-nomain STATIC ${CUKE_SOURCES})
-add_library(cucumber-cpp STATIC ${CUKE_SOURCES} main.cpp)
+if(CUKE_ENABLE_SHARED_LIB)
+  set(CUKE_LINKAGE SHARED)
+else()
+  set(CUKE_LINKAGE STATIC)
+endif()
+
+add_library(cucumber-cpp-nomain ${CUKE_LINKAGE} ${CUKE_SOURCES})
+add_library(cucumber-cpp ${CUKE_LINKAGE} ${CUKE_SOURCES} main.cpp)
```
Comment (ala-ableton) [inline src/CMakeLists.txt]: Same thing here, I don't think we should change the `CMakeLists.txt` files in this commit.

**Review by ala-ableton — COMMENTED** [review]
I only reviewed the first commit yet.

File: src/HookRegistrar.cpp [hunk 4]
```diff
@@ -95,7 +95,6 @@ void HookRegistrar::execAfterHooks(Scenario *scenario) {
     execHooks(afterHooks(), scenario);
 }
 
-
```
Comment (ala-ableton) [inline src/HookRegistrar.cpp]: Removing this line doesn't seem necessary.

File: src/Regex.cpp [hunk 5]
```diff
@@ -1,4 +1,5 @@
 #include <cucumber-cpp/internal/utils/Regex.hpp>
+#include <cucumber-cpp/internal/hook/HookRegistrar.hpp>
```
Comment (ala-ableton) [inline src/Regex.cpp]: Please sort alphabetically.

File: src/Regex.cpp [hunk 6]
```diff
@@ -26,7 +27,8 @@ boost::shared_ptr<RegexMatch> Regex::find(const std::string &expression) const {
 
 FindRegexMatch::FindRegexMatch(const boost::regex &regexImpl, const std::string &expression) {
     boost::cmatch matchResults;
-    regexMatched = boost::regex_search(expression.c_str(), matchResults, regexImpl, boost::regex_constants::match_extra);
+    regexMatched = boost::regex_search(expression.c_str(), matchResults, regexImpl, boost::regex_constants::match_extra)
+                   && HookRegistrar().execStepMatchingHook(matchResults);
```
Comment (ala-ableton) [inline src/Regex.cpp]: Why not indent with only one level of indentation?

**Review by ala-ableton — CHANGES_REQUESTED** [review]
The title of the second commit should start with a verb.

File: CMakeLists.txt (hunk 1 above)
Comment (nre-ableton) [reply CMakeLists.txt]: True, fixed

File: include/cucumber-cpp/internal/CukeEngineImpl.hpp (hunk 2 above)
Comment (nre-ableton) [reply include/cucumber-cpp/internal/CukeEngineImpl.hpp]: Fixed

File: src/CMakeLists.txt (hunk 3 above)
Comment (nre-ableton) [reply src/CMakeLists.txt]: Fixed

**Pull request closed without merging**